ceno_zkvm/src/instructions/riscv/div.rs

```diff
@@ -1,124 +1,348 @@
+                    // a_neg * (1 - b_neg) + (1 - a_neg) * b_neg
+                    (a_neg.clone() + b_neg.clone()) - (Expression::<E>::from(2) * a_neg * b_neg)
+                };
+                let negative_division = cb.flatten_expr(|| "neg_division", neg_div_expr)?;
+
+                // Check for signed division overflow: i32::MIN / -1
+                let is_dividend_max_negative = IsEqualConfig::construct_circuit(
+                    cb,
+                    || "is_dividend_max_negative",
+                    dividend.value(),
+                    (1u64 << (BIT_WIDTH - 1)).into(),
+                )?;
+                let is_divisor_minus_one = IsEqualConfig::construct_circuit(
+                    cb,
+                    || "is_divisor_minus_one",
+                    divisor.value(),
+                    ((1u64 << BIT_WIDTH) - 1).into(),
+                )?;
+                let is_signed_overflow = cb.flatten_expr(
+                    || "signed_division_overflow",
+                    is_dividend_max_negative.expr() * is_divisor_minus_one.expr(),
+                )?;
+
+                // For signed division overflow, dividend = -2^31 and divisor
+                // = -1, so that quotient = 2^31 would be required for proper
+                // arithmetic, which is too large for signed 32-bit values.  In
+                // this case, quotient and remainder are required to be set to
+                // -2^31 and 0 respectively.  These values are assured by the
+                // constraints
+                //
+                //   2^31 = divisor * quotient + remainder
+                //   0 <= |remainder| < |divisor|
+                //
+                // The second condition is the same inequality as required when
+                // there is no overflow, so no special handling is needed.  The
+                // first condition is only different from the proper value in
+                // the left side of the equality, which can be controlled by a
+                // conditional equality constraint using fixed dividend value
+                // +2^31 in the signed overflow case.
+                let div_rel_expr =
+                    quotient_signed.expr() * divisor_signed.expr() + remainder_signed.expr();
+                cb.condition_require_equal(
+                    || "signed_division_relation",
+                    is_signed_overflow.expr(),
+                    div_rel_expr,
+                    // overflow replacement dividend value, +2^31
+                    (1u64 << (BIT_WIDTH - 1)).into(),
+                    dividend_signed.expr(),
+                )?;
+
+                // Check required inequalities for remainder value; change sign
+                // for remainder and divisor so that checked inequality is the
+                // usual unsigned one, 0 <= remainder < divisor
+                let remainder_pos_orientation = (Expression::ONE
```

Comment:
```diff
diff --git a/ceno_zkvm/src/instructions/riscv/div.rs b/ceno_zkvm/src/instructions/riscv/div.rs
index 0128a462..576d19db 100644
--- a/ceno_zkvm/src/instructions/riscv/div.rs
+++ b/ceno_zkvm/src/instructions/riscv/div.rs
@@ -248,12 +248,10 @@ impl<E: ExtensionField, I: RIVInstruction> Instruction<E> for ArithInstruction<E
                 // Check required inequalities for remainder value; change sign
                 // for remainder and divisor so that checked inequality is the
                 // usual unsigned one, 0 <= remainder < divisor
-                let remainder_pos_orientation = (Expression::ONE
-                    - Expression::<E>::from(2) * negative_division.expr())
-                    * remainder_signed.expr();
-                let divisor_pos_orientation = (Expression::ONE
-                    - Expression::<E>::from(2) * divisor_signed.is_negative.expr())
-                    * divisor_signed.expr();
+                let remainder_pos_orientation: Expression<E> =
+                    (1 - 2 * negative_division.expr()) * remainder_signed.expr();
+                let divisor_pos_orientation =
+                    (1 - 2 * divisor_signed.is_negative.expr()) * divisor_signed.expr();
```